### PR TITLE
feat(unparser): adding alias for table scan filter in sql unparser

### LINKED ISSUE #12368: Unparsing TableScan with an alias and a pushdown filter to SQL string
### Is your feature request related to a problem or challenge?

In https://github.com/apache/datafusion/pull/12158 , we support unparse a TableScan with some pushdown options(e.g. projection, filter, or fetch ...). However, there are some issues with unparsing a TableScan with an alias and a filter. 
Consider the following case:
```rust
    let table_scan_with_filter_alias = table_scan_with_filters(
        Some("t1"),  // projection
        &schema, // schema
        None,  // alias
        vec![col("id").gt(col("age"))],  // filter
    )?.alias("ta")?.build()?;
```
The implementation of https://github.com/apache/datafusion/pull/12158 will produce a SQL like
```sql
SELECT * FROM t1 AS ta WHERE (t1.id > t1.age)
```
but the SQL should be
```SQL
SELECT * FROM t1 AS ta WHERE (ta.id > ta.age)
```

We need also to process the expression of the filter.

### Describe the solution you'd like

Rewrite the columns used by the filter. Make their relation prefix be the alias instead of the table name.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #12368

### CODE REVIEW

**Review by dmitrybugakov — APPROVED** [review]

File: datafusion/sql/src/unparser/plan.rs [hunk 1]
```diff
@@ -587,12 +587,24 @@ impl Unparser<'_> {
...
+                        if let Some(ref mut rewriter) = filter_alias_rewriter {
+                            expr.rewrite(rewriter).data()
+                        } else {
+                            Ok(expr)
+                        }
+                    })
+                    .reduce(|acc, expr_result| match (acc, expr_result) {
+                        (Ok(acc_expr), Ok(expr)) => Ok(acc_expr.and(expr)),
+                        (Err(e), _) | (_, Err(e)) => Err(e),
+                    })
```
Comment (goldmedal) [inline datafusion/sql/src/unparser/plan.rs:604]: We can simplify here like this
```suggestion
                    .reduce(|acc, expr_result| {
                        acc.and_then(|acc_expr| {
                            expr_result.map(|expr| acc_expr.and(expr))
                        })
                    })
```

File: datafusion/sql/src/unparser/rewrite.rs [hunk 2]
```diff
@@ -300,3 +301,27 @@ fn find_projection(logical_plan: &LogicalPlan) -> Option<&Projection> {
         _ => None,
     }
 }
+
+pub struct TableAliasRewriter {
```
Comment (goldmedal) [inline datafusion/sql/src/unparser/rewrite.rs:305]: How about adding some doc for this public struct?

**Review by goldmedal — APPROVED** [review]
Thanks @Lordworms working on this. Overall looks good to me. Just leave some suggestions.

**Comment by alamb** [comment]
I think @goldmedal 's comments are all great. Once those have been addressed this PR will be ready to go from my persective. 

Thank you @Lordworms @goldmedal and @dmitrybugakov

File: datafusion/sql/src/unparser/rewrite.rs (hunk 2 above)
Comment (Lordworms) [reply datafusion/sql/src/unparser/rewrite.rs:305]: Sure

**Comment by Lordworms** [comment]
@goldmedal  @alamb  @dmitrybugakov  Thanks for the review!

**Review by alamb — APPROVED** [review]
Thanks -- looks good to me

**Pull request merged**